boards: sparkfun: Add support for micromod_nrf52840 board and

jeronimoagullo opened the pull request:
Initial support of Sparkfun Micromod ecosystem with micromod nrf52840 board, asset tracker shield and micromod header definition.

The idea of this commit is to set up the basis of the Sparkfun Micromod ecosystem, which is a standard based on M.2 connectors and a specification to increase the availability of parts (boards, sensors and carriers). Since this ecosystem suits perfectly the Zephyr philosophy, it would be great to support these boards.

**implemented and tested peripherals**
- UART bus connected to SARA-R5 modem with the `samples/net/cellular_modem` app
- QSPI with w25q128jv flash memory with the `samples/drivers/spi_flash` app
- SPI connected to a microSD with the `samples/subsys/fs/fs_sample` app
- I2C connected to the faul gauge max17048 with the `samples/fuel_gauge/max17048` app
- USB for serial debugging with the `samples/usb/console` app

**Resources**
Micromod ecosystem pin/peripherals definition: https://learn.sparkfun.com/tutorials/getting-started-with-micromod
Micromod nrf52840 board: https://www.sparkfun.com/products/16984
Micromod asset tracker carrier: https://www.sparkfun.com/products/17272

jeronimoagullo committed:
boards: sparkfun: Add micromod nrf52840 board and asset tracker shield

Initial support of Sparkfun Micromod ecosystem with micromod nrf52840
board, asset tracker shield and micromod header definition.

Signed-off-by: Jeronimo Agullo <jeronimoagullo97@gmail.com>

nordicjm commented on boards/sparkfun/micromod_nrf52840/board.yml:
```diff
@@ -0,0 +1,5 @@
+board:
+  name: micromod_nrf52840
```
should not have nrf52840 in the name otherwise you have to build using `-DBOARD=micromod_nrf52840/nrf52840`. Then update filenames to remove the nrf52840 part

jeronimoagullo replied:
The issue is that the board is named "MicroMod nRF52840 Processor" since "Micromod" is just the pin standard interface (like Arduino, feather, etc.). https://www.sparkfun.com/products/16984

Thus, using just "MicroMod" would not be appropriate since there are more Micromod board with different processors: https://www.sparkfun.com/micromod#processor_boards

So, how should I rename the board?

nordicjm replied:
That's what the `socs` field is used for, the board would just be `micromod` then you select which soc to build for, that could be `nrf52840` or `sam5u4` or any other CPU

jeronimoagullo replied:
All rigth! I was thinking about it. Nice, I will modify it and force a new push. Thanks!

nordicjm commented on boards/sparkfun/micromod_nrf52840/Kconfig.defconfig:
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2024 Jerónimo Agulló <jeronimoagullo97@gmail.com>
+# SPDX-License-Identifier: Apache-2.0
+
+if BOARD_MICROMOD_NRF52840
```
likewise this becomes `if BOARD_MICROMOD`

jeronimoagullo replied:
Ok

nordicjm commented on boards/sparkfun/micromod_nrf52840/Kconfig.micromod_nrf52840:
```diff
@@ -0,0 +1,7 @@
+# Copyright (c) 2024 Jerónimo Agulló <jeronimoagullo97@gmail.com>
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_MICROMOD_NRF52840
```
and here, etc.

jeronimoagullo replied:
Ok

nordicjm commented on boards/sparkfun/micromod_nrf52840/micromod_nrf52840.dts:
```diff
@@ -0,0 +1,198 @@
+			label = "image-1";
+			reg = <0x00073000 0x00067000>;
+		};
+		scratch_partition: partition@da000 {
+			label = "image-scratch";
+			reg = <0x000da000 0x0001e000>;
+		};
```
we don't use scratch partitions anymore

jeronimoagullo replied:
Ok, I have removed it.

nordicjm reviewed (changes requested):
Also lacking board documentation

jeronimoagullo commented:
I have updated the board files to suit the new hardware model v2. Thus, the board is "micromod" and the variant is "micromod/nrf52840". Besides, I have added the documentation for both the board and the shield.

nordicjm commented on boards/shields/sparkfun_carrier_asset_tracker/doc/SparkFun_MicroMod_Asset_Tracker_shield.png:
Convert to webp and put through https://tinypng.com/ please to reduce size

jeronimoagullo replied:
OK

nordicjm commented on boards/sparkfun/micromod/doc/img/sparkfun_micromod.png:
As above

jeronimoagullo replied:
OK

nordicjm commented on boards/shields/sparkfun_carrier_asset_tracker/doc/index.rst:
```diff
@@ -0,0 +1,118 @@
+| SPI_CS                | General purpose pin             |
++-----------------------+---------------------------------+
+
+Adetail definition of the Micromod standard can be found in the
```
A detailed? **on** the micromod website?

nordicjm commented on boards/shields/sparkfun_carrier_asset_tracker/doc/index.rst:
```diff
@@ -0,0 +1,118 @@
+Asset Tracke guide website`_.
+
+Pins Assignment of Sparkfun SparkFun MicroMod Asset Tracker Shield
+=======================================================
```
length of heading should match text

jeronimoagullo replied:
OK

nordicjm commented on boards/shields/sparkfun_carrier_asset_tracker/doc/index.rst:
```diff
@@ -0,0 +1,118 @@
+.. _semtech_sx1276mb1mas:
+
+Sparkfun SparkFun MicroMod Asset Tracker Shield
+################################
```
as above

jeronimoagullo replied:
OK

nordicjm commented on boards/shields/sparkfun_carrier_asset_tracker/sparkfun_carrier_asset_tracker.overlay:
```diff
@@ -0,0 +1,50 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/{
```
```suggestion
/ {
```

jeronimoagullo replied:
OK

nordicjm commented on boards/shields/sparkfun_carrier_asset_tracker/sparkfun_carrier_asset_tracker.overlay:
```diff
@@ -0,0 +1,50 @@
+
+/{
+	aliases {
+		modem-uart= &micromod_1_uart;
```
```suggestion
		modem-uart = &micromod_1_uart;
```

jeronimoagullo replied:
OK

nordicjm commented on boards/shields/sparkfun_carrier_asset_tracker/sparkfun_carrier_asset_tracker.overlay:
```diff
@@ -0,0 +1,50 @@
+                reg = <0>;
+                status = "okay";
+                mmc {
+                    compatible = "zephyr,sdmmc-disk";
+                    status = "okay";
```
tabs only for indent

jeronimoagullo replied:
OK

nordicjm commented on boards/sparkfun/micromod/doc/index.rst:
```diff
@@ -0,0 +1,202 @@
+.. _boardname_linkname:
+
+SparkFun MicroMod board Processor
+#############
```
As above

nordicjm commented on boards/sparkfun/micromod/doc/index.rst:
```diff
@@ -0,0 +1,202 @@
+- ``micromod/nrf52840``
+
+SparkFun MicroMod nRF52840 board Processor
+#############
```
As above

nordicjm commented on boards/sparkfun/micromod/doc/index.rst:
```diff
@@ -0,0 +1,202 @@
+      CONFIG_SHELL_BACKEND_RTT=y
+
+Testing the 128Mb qspi memory flash
+************************************
```
as above

nordicjm commented on boards/sparkfun/micromod/doc/index.rst:
```diff
@@ -0,0 +1,202 @@
+
+Testing the 128Mb qspi memory flash
+************************************
+The ``micromod/nrf52840`` board variant has a built-in NOR flash memory connected
```
this is not a variant

jeronimoagullo replied:
I change it to SoC since it is just the same board with different SoC

nordicjm commented on boards/sparkfun/micromod/doc/index.rst:
```diff
@@ -0,0 +1,202 @@
+Testing the 128Mb qspi memory flash
+************************************
+The ``micromod/nrf52840`` board variant has a built-in NOR flash memory connected
+to the qspi interface. It can be tested with the samples/drivers/spi_flash app.
+
+Add the config overlay file to the boards directory of the sample
+
+   .. code-block::
+      CONFIG_NORDIC_QSPI_NOR=y
+      CONFIG_NORDIC_QSPI_NOR_FLASH_LAYOUT_PAGE_SIZE=4096
+      CONFIG_NORDIC_QSPI_NOR_STACK_WRITE_BUFFER_SIZE=16
+
+Then build and flash the application in the usual way.
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/drivers/spi_flash
+   :board: micromod/nrf52840
+   :goals: build flash
```
@de-nordic is this something we want?

nordicjm commented on boards/sparkfun/micromod/micromod_nrf52840_defconfig:
```diff
@@ -0,0 +1,22 @@
+# Enable RTT
+CONFIG_USE_SEGGER_RTT=y
+
+# enable GPIO
+CONFIG_GPIO=y
+
+# enable I2C
+CONFIG_I2C=y
+
+# enable SPI
+CONFIG_SPI=y
+
+# enable uart
```
*Enable (capital)

jeronimoagullo replied:
OK

nordicjm commented on boards/sparkfun/micromod/micromod_nrf52840_defconfig:
```diff
@@ -0,0 +1,22 @@
+# Enable RTT
+CONFIG_USE_SEGGER_RTT=y
+
+# enable GPIO
+CONFIG_GPIO=y
+
+# enable I2C
+CONFIG_I2C=y
+
+# enable SPI
+CONFIG_SPI=y
+
+# enable uart
```
*UART

jeronimoagullo replied:
OK

jeronimoagullo commented:
Hi! I would like to know the state of the review, is there something else that I need to carry out? Thanks in advance

kartben commented on boards/shields/sparkfun_carrier_asset_tracker/doc/SparkFun_MicroMod_Asset_Tracker_shield.webp:
move to `.../doc/img/sparkfun_carrier_asset_tracker.webp`

jeronimoagullo replied:
ok!

kartben replied:
this hasn't been addressed

kartben commented on boards/shields/sparkfun_carrier_asset_tracker/doc/index.rst:
```diff
@@ -0,0 +1,118 @@
+.. _semtech_sx1276mb1mas:
```
copy paste error / this needs to be unique to this shield

jeronimoagullo replied:
you are right, done!

kartben commented on boards/shields/sparkfun_carrier_asset_tracker/doc/index.rst:
```diff
@@ -0,0 +1,118 @@
+Pins Assignment of Sparkfun SparkFun MicroMod Asset Tracker Shield
+==================================================================
+
+The SparkFun MicroMod Asset Tracker Carrier Shield uses a 76 pins M.2
+connector. The following table depicts the interfaces and pin supported
+by Zephyr:
```
```suggestion
The SparkFun MicroMod Asset Tracker Carrier Shield uses a 76-pin M.2
connector. The following table depicts the interfaces and pins supported:
```

jeronimoagullo replied:
done!

kartben replied:
nope? :)

kartben replied:
@jeronimoagullo still not fixed?

kartben commented on boards/sparkfun/micromod/doc/index.rst:
```diff
@@ -0,0 +1,202 @@
+and can insert into any MicroMod carrier board. More information can be found in
+the `Micromod specification website`_.
+
+All Micromod board configurations support the following hardware features:
+- USB host mode compliant to 2.0 specification
```
```suggestion
All Micromod board configurations support the following hardware features:

- USB host mode compliant to 2.0 specification
```

jeronimoagullo replied:
done!

kartben commented on boards/sparkfun/micromod/doc/index.rst:
```diff
@@ -0,0 +1,202 @@
+
+   Sparkfun Micromod board (Credit: https://www.sparkfun.com)
+
+Zephyr currently supports the following SoCs:
+- ``micromod/nrf52840``
```
```suggestion
Zephyr currently supports the following SoCs:

- ``micromod/nrf52840``
```

jeronimoagullo replied:
done!

kartben commented on boards/sparkfun/micromod/doc/index.rst:
```diff
@@ -0,0 +1,202 @@
+Connections and IOs
+===================
+
+LED
+---
+* Led0 (blue) = P0.13
```
```suggestion
LED
---

* Led0 (blue) = P0.13
```

kartben commented on boards/sparkfun/micromod/doc/index.rst:
```diff
@@ -0,0 +1,202 @@
+Refer to the :ref:`nordic_segger` page to learn about debugging Nordic boards with a
+Segger IC.
+
+In case of using the jlink for debugging, you may want to use RTT viewer for displaying
+debug messages. In this case, you need to add the following configuration into your
+prj.conf file:
+
+   .. code-block::
+      # Enable RTT
+      CONFIG_CONSOLE=y
+      CONFIG_UART_CONSOLE=n
+      CONFIG_RTT_CONSOLE=y
+      CONFIG_USE_SEGGER_RTT=y
+      CONFIG_SHELL=y
+      CONFIG_SHELL_BACKENDS=y
+      CONFIG_SHELL_BACKEND_RTT=y
```
maybe just refer to the RTT snippet?
https://docs.zephyrproject.org/latest/snippets/rtt-console/README.html

jeronimoagullo replied:
@kartben Ohh I didn't hear before about snippets. It is exactly what I needed to avoid copying this configuration settings in each project. Definitely yes, I will refer to the RTT snippet.

However, I have never added documentation before in Zephyr, should I refer to RTT snippet using "target-notes::" as any other link or is there an internal way of doing it (for Zephyr documentation I mean)?

jeronimoagullo replied:
Nice! thank you very much for the explanation

kartben commented on boards/sparkfun/micromod/doc/index.rst:
```diff
@@ -0,0 +1,202 @@
+debug messages. In this case, you need to add the following configuration into your
+prj.conf file:
+
+   .. code-block::
```
```suggestion
   .. code-block:: cfg
```

kartben commented on boards/sparkfun/micromod/doc/index.rst:
```diff
@@ -0,0 +1,202 @@
+
+Add the config overlay file to the boards directory of the sample
+
+   .. code-block::
```
```suggestion
   .. code-block:: cfg
```

kartben commented on boards/shields/sparkfun_carrier_asset_tracker/doc/index.rst:
```diff
@@ -0,0 +1,118 @@
+   Sparkfun SparkFun MicroMod Asset Tracker Shield (Credit: Sparkfun)
+
+More information about the shield can be found at the `SparkFun MicroMod
+Asset Tracke guide website`_.
```
```suggestion
Asset Tracker guide website`_.
```

kartben commented on boards/shields/sparkfun_carrier_asset_tracker/doc/index.rst:
```diff
@@ -0,0 +1,118 @@
+
+.. target-notes::
+
+.. _SparkFun MicroMod Asset Tracke guide website:
```
```suggestion
.. _SparkFun MicroMod Asset Tracker guide website:
```

kartben commented on boards/sparkfun/micromod/doc/index.rst:
```diff
@@ -0,0 +1,202 @@
+The ``micromod/nrf52840`` board configuration features the nRF52840 SoC
+from Nordic Semiconductor, the SparkFun MicroMod nRF52840 Processor offers
+a powerful combination of ARM Cortex-M4 CPU and 2.4 GHz Bluetooth transceiver
+in the MicroMod form-factor. More informmation can be found in
```
```suggestion
in the MicroMod form-factor. More information can be found in
```

kartben commented on boards/shields/sparkfun_carrier_asset_tracker/sparkfun_carrier_asset_tracker.overlay:
```diff
@@ -0,0 +1,50 @@
+	};
+};
+
+&micromod_0_i2c{
```
```suggestion
&micromod_0_i2c {
```

jeronimoagullo commented:
Hi! Since the new release will be frozen in two weeks... Is there something else that I should modify to get this pull request ready? Thanks in advance

nordicjm commented on boards/sparkfun/micromod/doc/index.rst:
```diff
@@ -0,0 +1,212 @@
+Supported Features
+==================
+
+The Zephyr ``micromod/nrf52840`` board configuration supports the following hardware
```
board target

nordicjm commented on boards/sparkfun/micromod/doc/index.rst:
```diff
@@ -0,0 +1,212 @@
+Overview
+********
+
+The ``micromod/nrf52840`` board configuration features the nRF52840 SoC
```
board target

nordicjm commented on boards/sparkfun/micromod/doc/index.rst:
```diff
@@ -0,0 +1,212 @@
+Testing the 128Mb qspi memory flash
+***********************************
+
+The ``micromod/nrf52840`` board SoC has a built-in NOR flash memory connected
```
board target, there is no spi nor built into the soc

nordicjm commented on boards/sparkfun/micromod/micromod_nrf52840.dts:
```diff
@@ -0,0 +1,197 @@
+		status="okay";
+		compatible = "nordic,qspi-nor";
+		reg = <0>;
+		// Currently, it supports only pp mode
+		writeoc = "pp";
+		// Currently, it supports only fastread and read2o
+		readoc = "fastread";
```
comments don't look accurate, datasheet for `w25q128jv` shows it supports spi/dspi/qspi

nordicjm commented on boards/sparkfun/micromod/micromod_nrf52840.dts:
```diff
@@ -0,0 +1,197 @@
+		writeoc = "pp";
+		// Currently, it supports only fastread and read2o
+		readoc = "fastread";
+		sck-frequency = <DT_FREQ_M(8)>;
```
you can run this much faster, 8MHz is the max for the mx25 on the nrf52840dk because that has an ultra-low power mode which limits the speed to 8MHz but the `w25q128jv` datasheet shows it supports much faster speeds (optional change)

nordicjm commented on boards/sparkfun/micromod/micromod_nrf52840.dts:
```diff
@@ -0,0 +1,197 @@
+			label = "mcuboot";
+			reg = <0x00000000 0x0000C000>;
+		};
+		slot0_partition: partition@c000 {
+			label = "image-0";
+			reg = <0x0000C000 0x00076000>;
+		};
+
+		slot1_partition: partition@82000 {
+			label = "image-1";
+			reg = <0x00082000 0x00076000>;
+		};
```
if using swap using move then these sizes are not optimal, if you build mcuboot with debug logging then it will show a message explaining (optional)

jeronimoagullo replied:
I used to use scratch instead of moving, I have read about it and I fixed it increasing primary slot. Now, the mentioned debug log is not displaying. Thanks!

nordicjm commented on boards/sparkfun/micromod/micromod_nrf52840_defconfig:
```diff
@@ -0,0 +1,22 @@
+# Enable RTT
+CONFIG_USE_SEGGER_RTT=y
+
+# enable GPIO
+CONFIG_GPIO=y
+
+# enable I2C
+CONFIG_I2C=y
+
+# enable SPI
+CONFIG_SPI=y
```
*Enable (capital E)

nordicjm commented on boards/sparkfun/micromod/micromod_nrf52840_defconfig:
```diff
@@ -0,0 +1,22 @@
+# enable GPIO
+CONFIG_GPIO=y
+
+# enable I2C
+CONFIG_I2C=y
+
+# enable SPI
+CONFIG_SPI=y
```
Why is i2c and spi enabled by default?

jeronimoagullo replied:
I have removed them, I forgot to eliminate from here.

nordicjm commented on boards/sparkfun/micromod/micromod_nrf52840.dts:
```diff
@@ -0,0 +1,196 @@
+		reg = <0>;
+		writeoc = "pp4io";
+		readoc = "read4io";
+		sck-frequency = <DT_FREQ_M(50)>;
```
32MHz is max QSPI speed on nrf52840

nordicjm commented on boards/sparkfun/micromod/micromod_nrf52840_defconfig:
```diff
@@ -0,0 +1,19 @@
+# Enable RTT
+CONFIG_USE_SEGGER_RTT=y
+
+# enable GPIO
+CONFIG_GPIO=y
+
+# enable console
```
*Enable (with capital)

jeronimoagullo replied:
Sorry, I fix it

nordicjm reviewed (dismissed):
Looks good, thanks, 2 minor nits.

kartben commented on boards/sparkfun/micromod/doc/index.rst:
```diff
@@ -0,0 +1,212 @@
+Micromod header
+---------------
+
+* micromod_1_uart = uart0
+* micromod_2_uart = uart1
```
fwiw it's a bit confusing that this starts at 1 while i2c and spi start at 0

jeronimoagullo replied:
I fully agree with you. However, it is the naming that Sparkfun uses in the micromod series (UART1, UART2, I2C(0) and I2C1: https://cdn.sparkfun.com/assets/learn_tutorials/1/2/0/6/SparkFun_MicroMod_Interface_v1.0_-_Pin_Descriptions.pdf

What do you think that is better?

kartben replied:
fine then :) I was mainly asking in case this was an oversight

kartben commented on boards/sparkfun/micromod/doc/index.rst:
```diff
@@ -0,0 +1,212 @@
+
+Add the config overlay file to the boards directory of the sample
+
+   .. code-block:: kconfig
```
"kconfig" is for actual kconfig file, not ".config" files

```suggestion
   .. code-block:: cfg
```

jeronimoagullo replied:
I have modified this section a bit, because the best sample is the `/samples/drivers/jesd216` app instead of the `spi_flash` sample since it is a QSPI flash. I got the samples mixed. I checked it out and works with the jesd216 sample smoothly :D

kartben commented on boards/sparkfun/micromod/doc/index.rst:
```diff
@@ -0,0 +1,212 @@
+The ``micromod/nrf52840`` board target supports the following
+hardware features:
+
+- ARM Cortex-M4 CPU with floating point unit (FPU)
+   - 1MB internal Flash
+   - 256kB internal RAM
+- Integrated 2.4GHz radio with support for Bluetooth Low Energy (BLE) and ieee802154
```
```suggestion
- ARM Cortex-M4 CPU with floating point unit (FPU)

   - 1MB internal Flash
   - 256kB internal RAM

- Integrated 2.4GHz radio with support for Bluetooth Low Energy (BLE) and ieee802154
```

renders as a "definition list" otherwise

jeronimoagullo replied:
ok!

kartben reviewed (changes requested):
looks like some of my comments from earlier you marked as "done" didn't make it, in the current version of the PR at least. Spotted a few more doc issues. Thanks!

jeronimoagullo commented:
@kartben please, check out that the images of the shield and board are in the corresponding `doc/img` folder since I have not modified it in this last push, but I see them in the corresponding folders.

kartben commented on boards/shields/sparkfun_carrier_asset_tracker/doc/img/SparkFun_MicroMod_Asset_Tracker_shield.webp:
still named incorrectly :) should be `sparkfun_carrier_asset_tracker.webp` please

jeronimoagullo replied:
ohh I was focusing on the folder and I didn't notice that the issue was the name, sorry... It is fixed now :D

jeronimoagullo commented:
Hi! I would like to know the PR status since I would like to close it before v3.7.0 release. I have carried out the @kartben and @nordicjm request. Is there something else that I should modify? thanks!

kartben reviewed (dismissed):
docs page can probably be polished some more but good enough to get in IMO

kartben commented:
@jeronimoagullo you'll need to rebase as the CI failure is unrelated to your change and as since then been addressed

jeronimoagullo commented on boards/sparkfun/micromod/micromod_nrf52840.dts:
```diff
@@ -0,0 +1,196 @@
+	aliases {
+		led0 = &led0;
+		watchdog0 = &wdt0;
+		spi-flash0 = &w25q128jv;
```
Hi @anangl I didn't know it, I have removed this alias. Thanks!

jeronimoagullo commented:
I have added minor changes according to check failures. It should pass the test now:
- The nrf52840 cannot have two UARTS. Thus `uart0` is now "UARTE" and `uart1` is "UART" (similar to the nrf52840dk board).
- Besides, I have added an overlay into `tests/drivers/adc/adc_api` test including `"nordic,nrf-saadc-common.dtsi"`

jeronimoagullo commented:
Now it passes all checks @kartben @nordicjm  @anangl

anangl commented on boards/sparkfun/micromod/micromod_nrf52840.dts:
```diff
@@ -0,0 +1,200 @@
+};
+
+&uart1 {
+	compatible = "nordic,nrf-uart";
```
This is incorrect. `uart1` cannot be "nordic,nrf-uart", only "nordic,nrf-uarte" (see [UART instances](https://docs.nordicsemi.com/bundle/ps_nrf52840/page/uart.html#d1445e350) and [UARTE instances](https://docs.nordicsemi.com/bundle/ps_nrf52840/page/uarte.html#d1474e397) in nRF52840).

jeronimoagullo replied:
you are rigth, I will remove this line since it is defined as UARTE in the `nrf52840.dtsi`

jeronimoagullo replied:
@anangl done!

nordicjm commented on boards/sparkfun/micromod/micromod_nrf52840.dts:
```diff
@@ -0,0 +1,199 @@
+};
+
+&uart1 {
+	current-speed = <115200>;
```
this is missing
```
	compatible = "nordic,nrf-uarte";
```

jeronimoagullo commented:
Hi @anangl it seems to be ready to merge
